Fix GDScript base and outer classes, signals and functions lookup order

adamscott opened the pull request:
Fix GDScript inner and outer classes lookup order
- Add outer class lookup test
- Add signal lookup test

Fixes #70216 (again)
Fixes #70495 - Accessing keys() dict of an enum defined in a singleton crashes app

adamscott committed:
Fix GDScript base and outer classes, signals and functions lookup order

- Add outer class lookup test
- Add signal lookup test

Co-authored-by: Dmitrii Maganov <vonagam@gmail.com>

vonagam commented:
Hah, but unfortunately not right. There is difference in logic for base and outer classes. In base classes they do handle variables, functions and signals, while in outer lookup it is ignored.

So, the next example based on that:
 ```gdscript
 class A:
  var Q := 'right one'

class X:
  const Q := 'wrong one'

class Y extends X:
  class B extends A:
    func check() -> void:
      print(Q)

func _ready() -> void:
  var b := Y.B.new()
  b.check()
 ```

adamscott commented:
> Hah, but unfortunately not right. There is difference in logic for base and outer classes. In base classes they do handle variables, functions and signals, while in outer lookup it is ignored.
> 
> So, the next example based on that:
> 
> ```gdscript
> class A:
>  var Q := 'right one'
> 
> class X:
>  const Q := 'wrong one'
> 
> class Y extends X:
>  class B extends A:
>    func check() -> void:
>      print(Q)
> 
> func _ready() -> void:
>  var b := Y.B.new()
>  b.check()
> ```

Please. I do please you to create a complete example in advance.

vonagam commented:
Yeah, sorry for that, my reasoning is that I try to show the simplest example that everyone will see is wrong, not the most convoluted one (because it will be confusing and less in priority).

When you present a solution, I look at it and analyze what would be the next simplest showcase that the solution is wrong.

I also pointed out the root of the problem - as I wrote in the issue opening text you cannot continue using `get_class_node_current_scope_classes` just for outer logic. So when the known root of the problem is not addressed it is quite easy for me to come up with next example.

adamscott commented:
> When you present a solution, I look at it and analyze what would be the next simplest showcase that the solution is wrong.

The problem with this is that I'm stuck at doing ping-pong.

> as I wrote in the issue opening text you cannot continue using `get_class_node_current_scope_classes` just for outer logic.

No. You said that it wasn't working because:

> [...] but it does not because of the wrong usage of `get_class_node_current_scope_classes` every class connected in any way to outer class gets checked even before first base class of the type in question is looked at.

... which I corrected. And again, I just updated the PR. `get_class_node_current_scope_classes()` work wonderfully, it was just missing what to do when encountering variable, functions and signals, as you said.

> I also pointed out the root of the problem

Please, when "pointing out the root of a problem", please add some context or a solution yourself. And especially don't say that it's easy.

vonagam commented:
> please add some context or a solution yourself

I think I provided the context, but solution is not a one-line change. My thoughts: I think that `get_class_node_current_scope_classes` has value and to be consistent with other places it should be used to replace whole `while (base_class != nullptr)` loop and in iteration you need to check if the current script is outer or not, if it is base one then instance things like variables/functions/signals should be accepted, otherwise not.

adamscott commented:
@vonagam I replaced the list value of `GDScriptParser::ClassNode *` with a new list of `GDScriptAnalyzer::CurrentScopeClass` structs. It adds metadata, as you suggested, to know if the current class node "type" (I'll probably rename it) is `BASE` or `OUTER`.

Then, I realized that `get_class_node_current_scope_classes()` in `reduce_identifier_from_base()` already loops through the base, so I removed the `while (base != nullptr)` to introduce a single `if (base != nullptr)` condition.

@rune-scape @anvilfolk If you could add your input to this complex issue, it would be greatly appreciated.

anvilfolk commented:
Been looking at this a little bit, and I'm starting to wonder if maybe we shouldn't make the analyzer a little more flexible. What if we had a bitfield in `reduce_identifier_from_base` and the like, which defines which types of things are in scope.

I can imagine this bitfield having:

- `CLASS_TYPES`, such as inner classes,
- `CLASS_MEMBER`, such as static functions & enums
- `INSTANCE_MEMBER` such as variables and non-static functions), and `ALL` for everything.
- `ALL` all of the above

This way, depending on what we are searching for (e.g. whether we're looking into base classes or outer classes), we could pass the method some information stating that it should ignore instance members and class types, and only look for class members.

Would this make sense at all?

vonagam commented:
Alternative solution to separate bases from outer:
`get_class_node_current_scope_classes` currently returns base classes first then outer ones. So if you get a number of base classes either by returning or passing int by reference then by comparing if current index is less than a number of base classes or not you will know it's type. That's an option that requires no additional types/structures, but just an option.

adamscott commented:
> `get_class_node_current_scope_classes` currently returns base classes first then outer ones.

I think you misunderstand the method.

It returns, yes, base classes in priority over outer classes. But the recursive nature of the method makes it so that outer classes of the first base class have priority on the second base class.

adamscott commented:
> It returns, yes, base classes in priority over outer classes. But the recursive nature of the method makes it so that outer classes of the first base class have priority over the second base class.

Whoa, I'm totally wrong. You're right!

adamscott commented:
Just found out how to add support for signals. It's in the PR.

anvilfolk commented:
Did you get rid of the `OUTER` and `BASE` struct thing? I love that a looooooot! I was wondering whether there wasn't a way to do it without that extra complication :)

adamscott commented:
> Did you get rid of the `OUTER` and `BASE` struct thing? I love that a looooooot! I was wondering whether there wasn't a way to do it without that extra complication :)

It was @vonagam's idea!

vonagam commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3112,43 +3116,64 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 
-			GDScriptParser::ClassNode::Member member = base_class->get_member(name);
+			GDScriptParser::ClassNode::Member member = script_class->get_member(name);
 			p_identifier->set_datatype(member.get_datatype());
```
This needs to be moved to individual cases, because the member might be of irrelevant type (like an instance variable of an outer class).

vonagam commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3157,49 +3182,16 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
+				} break;
+
 				default:
 					break; // Type already set.
```
This comment will be outdated after update, can be removed.

adamscott commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3113,8 +3117,7 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
+	if (base_class != nullptr) {
 		if (base_class->identifier && base_class->identifier->name == name) {
 			p_identifier->set_datatype(base_class->get_datatype());
 			return;
```
I tested this and it did not seem to fix the issue you mentionned.

vonagam replied:
To fix it you need to add `p_base == nullptr` check. It does not make sense to type access to the class names through class itself (`A.A` or `B.A`) or class instance (`a.A` or `b.A`) (this will lead to runtime error anyway). Only when there is no base (so simply `A`) it can be typed properly. And it should have your constant logic so that you can do `const X := A` inside class itself, for alias for example.

vonagam commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3112,43 +3116,64 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
+				case GDScriptParser::ClassNode::Member::CONSTANT: {
 					p_identifier->is_constant = true;
 					p_identifier->reduced_value = member.constant->initializer->reduced_value;
 					p_identifier->source = GDScriptParser::IdentifierNode::MEMBER_CONSTANT;
```
I think it is. It was mistake before not to set it in outers.

vonagam commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3112,94 +3116,95 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
+				}
+
+				case GDScriptParser::ClassNode::Member::CLASS: {
 					if (p_base != nullptr && p_base->is_constant) {
```
This condition should be replace with `p_base == nullptr || base.is_meta_type`. So:
```gdscript
class Outer:
  class Inner: pass
  const Alias0 := Inner # p_base == nullptr
const Alias1 := Outer.Inner # base.is_meta_type
```

anvilfolk commented:
I like the latest changes! Code is getting super nicely refactored!!!

I wonder if the new function should be called `reduce_identifier_from_base`, or maybe something slightly different like `set_identifier_to_script_class()`? when I looked in there I was expecting it to be doing a search through the base class members, but that's not it at all :)

vonagam commented:
Another important thing:
```gdscript
class A:
  var x := 3

func _ready() -> void:
  print(A.x) # no warning right now
```
Before `reduce_identifier_from_base` was used only on instances, but now it is used on classes too. So those cases need to be distinguished. Since `is_base` used for those relevant things, you can rename it to `is_base_instance` and initialize it with `!type.is_meta_type`.

adamscott commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3114,102 +3130,88 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
+					p_identifier->set_datatype(member.get_datatype());
+					if (base.is_constant) {
 						p_identifier->is_constant = true;
 						p_identifier->source = GDScriptParser::IdentifierNode::MEMBER_CONSTANT;
```
@vonagam I rebased the PR and used `base.is_constant` instead of `p_base->is_constant`.

vonagam replied:
Oh. Two notes.

a) I would have preferred `base.is_meta_type` as more self-explanatory than `base.is_constant`.
But since there is no possibility for a constant instance at the moment it does not matter much.

b) After seeing this check though, it seems wrong to have. It is the same deal that we had with a nested classes. Since all static things are exposed through an instance - constants, enum values, nested classes, static functions - it is strange not to expose an enum too. And the fact that there is a call to `set_datatype` before `if` means that it is exposed. So I checked:
```gdscript
class A:
  enum Hello { World }

func _ready() -> void:
  var a := A.new()
  print(a.Hello) # { "World": 0 }
  print(A.Hello) # <null>
```
I think `if` guard should be removed and `reduced_value` needs to be set.

adamscott commented:
@anvilfolk @rune-scape If you could take a final look to the PR, we think, @vonagam and I, that the PR is ready to merge.

Lippanon commented:
This PR fixes the crash when running the project whose backtrace I posted in https://github.com/godotengine/godot/issues/70495.
However it still shows many errors when starting:
![image](https://user-images.githubusercontent.com/115422160/209447448-e5c71ef9-beca-498e-be11-3e856d595f67.png)

anvilfolk commented:
Just looked through the code, here's my understanding of what it does:

`reduce_identifier_from_base` now keeps track of whether the base is an instance or whether the base is a type (using `is_base_instance` and through `is_meta_type`. If the base is an instance, then you are free to check base & outer classes for instance variables. If the base is not an instance, then you check only for class members such as static functions, enums, constants, etc.

This generally makes sense, but I can't tell why this wouldn't search for instance members of an outer class if the base is an instance.

So something like

```GDScript
class_name Outer
var x

class Inner:
  func test():
    print(x) # check outer instance member from instance

static func test():
  var i = Inner.new()
  i.test() # Call on an instance
```


I'm compiling right now to test :)

anvilfolk commented:
Ok, all the tests I've come up with seem to work still, and I think I figured out why too :)

`get_class_node_current_scope_classes(base_class, &script_classes);` gets all the classes, prioritizing base classes first. At some point, you'll reach the outer class of the original base. HOWEVER!

Once that happens, `is_base_instance = is_base_instance && script_class->base_type.class_type != nullptr;` makes it so `is_base_instance = false`. That's because the topmost class in the class hierarchy has no `base_type.class_type`. Anything that happens after that is clamped to the base no longer being instance, meaning that after you have reached the top of the class hierarchy, you can only ever find static/class members, no longer instance members.

So that makes perfect sense, though maybe it does deserve a slightly longer comment about the order in which things come in `script_class`, because it's a pretty subtle interaction coming from two places.

But it makes sense and seems to work!

Though I can confirm that it does not make these errors go away, since I just reproduced them:
```core\variant\variant_call.cpp:1339 - Condition "!method" is true. Returning: false
core\variant\variant_call.cpp:1265 - Condition "!method" is true. Returning: 0
core\variant\variant_call.cpp:1318 - Condition "!method" is true. Returning: Variant::NIL
core\variant\variant_call.cpp:1251 - Condition "!method" is true. Returning: nullptr
```

Also, this line: `ERR_FAIL_COND_MSG(err != OK, "Error while getting full script.");` is triggering many many times, I think for scripts that might have errors in them, or not be fully parsed? Something about scripts with errors staying in the cache perhaps? I wonder if it isn't best not to have that error message?

anvilfolk reviewed (approved):
Approved after a fairly deep look. Looks very sensible and smart and lovely :) Would just love a little more commenting on the crucial line that makes `is_base_instance` false :)

anvilfolk commented:
Found out the core of the problem with the `!method` errors, at least as it relates to this PR.

This line in `reduce_identifier_from_base_set_class`:

`scr = scr->find_class(p_identifier_datatype.class_type->fqcn);`

Returns `null`. In my case, it was for this file:

```GDScript
extends Node
class_name MyClass # If it doesn't trigger for you, remove this line then add it again.

class B:
	pass

func _ready():
	var _test_instance = B.new()
```

It can't find `B` from its `fqcn`, which is `MyClass::B`. This results in the compiler attempting to compile it as a `BUILTIN` for some reason.

vonagam commented:
Right, `scr->find_class` can return null for nested classes and it can be reproduced by commenting/uncommenting them. That what was my comments about (like [here](https://github.com/godotengine/godot/pull/70055#issuecomment-1356299062)) in the #70055 that introduced that way to get `reduced_value` for classes.

This current PR does not changes the already existing logic, just moves it to `reduce_identifier_from_base_set_class` without any alteration.

The most likely solution would be what @rune-scape mentioned in [this comment](https://github.com/godotengine/godot/issues/69917#issuecomment-1350261984), so that you can request a cached script for a nested class directly by its class path (meaning not by getting a cached root script and only then searching for nested one in it), then cache will have opportunity to see that it is outdated and repopulate subscripts.

anvilfolk commented:
I did test it and it does change the logic slightly, since `find_class` was not used previously! :)

I did just submit a PR that fixes that potential regression though in #70537. Feel free leave a review there or just incorporate that in your PR though.

The issue is described there, and it's pretty subtle & very very localized to when you change `class_name`. It's not a fix for the other issues that we are seeing. It's good to know that you've found the issue. That's a good find, I had forgotten that whenever `is_constant` is set, then `reduced_value` also needs to be set!

adamscott commented:
@vnen What do you think of this PR?

adamscott commented:
@vnen @vonagam I rebased the PR with the merge of #70613.

I think the PR is ready.

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3024,6 +3028,18 @@ GDScriptParser::DataType GDScriptAnalyzer::make_global_class_meta_type(const Str
+
+	p_identifier->set_datatype(p_identifier_datatype);
+	Error err = OK;
+	GDScript *scr = GDScriptCache::get_full_script(p_identifier_datatype.script_path, err).ptr();
```
There's a bit of an issue about loading the script in the middle of the analyzer because this will not work if there are cyclic dependencies between the classes (that is, the script being loaded depends on the one being parsed, this will fail).

I'm not sure why this needs to be made a compile-time constant. I can see some benefits but I don't know if it can work properly in all cases so it might be better for the compiler to sort this out in the next phase.

adamscott replied:
> There's a bit of an issue about loading the script in the middle of the analyzer because this will not work if there are cyclic dependencies between the classes (that is, the script being loaded depends on the one being parsed, this will fail).

For now, the tests pass and no issue about this has been found yet. Do you have an idea of a test to add where the issue would happen?

> I'm not sure why this needs to be made a compile-time constant.

I added the inlining as it's just a set of instructions to execute, there's no logic beyond this function, like there's no need for this function other than reduce duplicated code. But I'm kinda a noob at C++, so I can remove the inline part.

vnen replied:
Looking at the code deeper now I understand why this "works". `get_full_script()` actually returns the shallow script here, since this is not asking to update from disk. The current script being loaded is already on the `full_gdscript_cache` because it 's added to the map **before** it's actually compiled, so the `get_full_script()` will never fail, but sometimes return a shallow (i.e. not compiled)  script instead.

So technically this will always work because even if it can't compile due to a reference cycle, it will still get the script that will be compiled later.

The main problem now is that you add a reference to the script in the function's list of constants, which will create a reference cycle that can never be cleared. Especially since this is also done for the very script being compiled now, so any reference to the same class will create this ref cycle (noted, it's not often that one references the same class by name, but there's no reason to keep a new reference to same script if that happens).

My question is still the same though: why does this need to be a compile-time constant? Even if you are accessing members from it, I believe the analyzer is able to retrieve the members directly without having to store the class. For instance: `var v = ClassName.InnerClass.SOME_CONST` can just grab `SOME_CONST` and store it directly as a constant. With this PR the `InnerClass` would also be stored in this case, which is unnecessary.

Unless I'm missing something else, making this constant does not serve any particular purpose and it's not needed in any case.

vonagam replied:
It started with need to make `const X := preload(...).X`. If subscript for an inner class is not constant then it cannot be assigned to a constant right now. And it is a constant then it is expected to have reduced_value.

vnen replied:
My issue is that with a script like this:

```gdscript
class_name Foo
const x = 0
func bar():
	print(Foo.x)
```

You now store a reference to the `Foo` script inside the constants of its own function `bar`. I understand storing references to other scripts (even if they make a cycle) but I don't see how this is necessary.

At the very least this should avoid storing the very same script being compiled.

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3336,7 +3341,8 @@ void GDScriptAnalyzer::reduce_identifier(GDScriptParser::IdentifierNode *p_ident
 
 	if (found_source) {
-		if ((p_identifier->source == GDScriptParser::IdentifierNode::MEMBER_VARIABLE || p_identifier->source == GDScriptParser::IdentifierNode::INHERITED_VARIABLE) && parser->current_function && parser->current_function->is_static) {
+		bool source_is_variable_or_signal = p_identifier->source == GDScriptParser::IdentifierNode::MEMBER_VARIABLE || p_identifier->source == GDScriptParser::IdentifierNode::INHERITED_VARIABLE || p_identifier->source == GDScriptParser::IdentifierNode::MEMBER_SIGNAL;
```
Since this adding signals, the error message should be chaged to reflect it, otherwise it might be confusing.

adamscott replied:
Fixed the error message!

vnen reviewed (commented):
The overall approach is good, I just have the one concerned I commented about loading scripts during parsing.

adamscott commented:
The build fails, but it's Windows that failed to build the artifacts, not that it had errors elsewhere.

vonagam reviewed (approved):
I'm coauthoring (for part inside `reduce_identifier_from_base`), so do approve.

Though, it is bugging me that `get_full_script` is used for getting a script for a class and not `get_shallow_script` with `raise_status` as it is done for the same purpose in preload. But at this point it is already a preexisting thing, this PR extends what this logic applies to but does not change it. I do hope that in future shallow script can be made to work.

Otherwise everything is perfect.

adamscott commented:
> and not `get_shallow_script` with `raise_status` as it is done for the same purpose in preload.

I did not try that, I'll try!

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -3164,18 +3169,20 @@ void GDScriptAnalyzer::reduce_identifier(GDScriptParser::IdentifierNode *p_ident
 		if (!lambda_stack.is_empty()) {
 			// If the identifier is a member variable (including the native class properties), we consider the lambda to be using `self`, so we keep a reference to the current instance.
-			if (p_identifier->source == GDScriptParser::IdentifierNode::MEMBER_VARIABLE || p_identifier->source == GDScriptParser::IdentifierNode::INHERITED_VARIABLE) {
+			if (source_is_variable || source_is_variable) {
```
```suggestion
			if (source_is_variable || source_is_signal) {
```
I think you meant this, right?

adamscott replied:
Done.

vnen reviewed (approved):
I believe we can merge as it is since it works and fixes the issues. I still have concerns about the lifetime of the scripts since I believe there are some cases creating reference cycles, but we can solve those problems (if they do exist indeed) properly later on.

akien-mga commented:
Thanks to both of you, huge work!
